src/ripple/app/ledger/impl/LedgerCleaner.cpp

```diff
@@ -96,20 +101,24 @@ class LedgerCleanerImp
     //
     //--------------------------------------------------------------------------
 
-    void onPrepare ()
+    void onPrepare () override
     {
     }
 
-    void onStart ()
+    void onStart () override
     {
-        startThread();
+        thread_ = std::thread {&LedgerCleanerImp::run, this};
     }
 
-    void onStop ()
+    void onStop () override
     {
-        m_journal.info << "Stopping";
-        signalThreadShouldExit();
-        notify();
+        JLOG (j_.info) << "Stopping";
+        {
+            std::lock_guard<std::mutex> lock (mutex_);
+            state_ = State::shouldExit;
+            wakeup_.notify_one();
```

Comment:
I agree - it's not a deadlock. But doesn't it still make sense to put the `notify_one` outside the lock? I'm picturing waking up a thread that isn't able to run, then releasing the lock, then waking up the thread again. (I also haven't messed with `condition_variable`, so this is new to me as well).